File: go/yamlstar_test.go
// Copyright 2024 yaml.org
// MIT License

package yamlstar_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/yaml/yamlstar/go"
)

func TestLoadSimpleScalar(t *testing.T) {
	data, err := yamlstar.Load("hello")
	require.NoError(t, err)
	assert.Equal(t, "hello", data)
}

func TestLoadInteger(t *testing.T) {
	data, err := yamlstar.Load("42")
	require.NoError(t, err)
	// JSON numbers are always float64 in Go
	assert.Equal(t, float64(42), data)
}

func TestLoadFloat(t *testing.T) {
	data, err := yamlstar.Load("3.14")
	require.NoError(t, err)
	assert.Equal(t, 3.14, data)
}

func TestLoadBooleanTrue(t *testing.T) {
	data, err := yamlstar.Load("true")
	require.NoError(t, err)
	assert.Equal(t, true, data)
}

func TestLoadBooleanFalse(t *testing.T) {
	data, err := yamlstar.Load("false")
	require.NoError(t, err)
	assert.Equal(t, false, data)
}

func TestLoadNull(t *testing.T) {
	data, err := yamlstar.Load("null")
	require.NoError(t, err)
	assert.Nil(t, data)
}

func TestLoadSimpleMapping(t *testing.T) {
	data, err := yamlstar.Load("key: value")
	require.NoError(t, err)
	assert.Equal(t, map[string]any{"key": "value"}, data)
}

func TestLoadNestedMapping(t *testing.T) {
	yaml := `
outer:
  inner: value
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)
	expected := map[string]any{
		"outer": map[string]any{
			"inner": "value",
		},
	}
	assert.Equal(t, expected, data)
}

func TestLoadMappingMultipleKeys(t *testing.T) {
	yaml := `
key1: value1
key2: value2
key3: value3
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)
	expected := map[string]any{
		"key1": "value1",
		"key2": "value2",
		"key3": "value3",
	}
	assert.Equal(t, expected, data)
}

func TestLoadSimpleSequence(t *testing.T) {
	yaml := `
- item1
- item2
- item3
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)
	assert.Equal(t, []any{"item1", "item2", "item3"}, data)
}

func TestLoadFlowSequence(t *testing.T) {
	data, err := yamlstar.Load("[a, b, c]")
	require.NoError(t, err)
	assert.Equal(t, []any{"a", "b", "c"}, data)
}

func TestLoadTypeCoercion(t *testing.T) {
	yaml := `
string: hello
integer: 42
float: 3.14
bool_true: true
bool_false: false
null_value: null
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)

	m, ok := data.(map[string]any)
	require.True(t, ok)

	assert.Equal(t, "hello", m["string"])
	assert.Equal(t, float64(42), m["integer"])
	assert.Equal(t, 3.14, m["float"])
	assert.Equal(t, true, m["bool_true"])
	assert.Equal(t, false, m["bool_false"])
	assert.Nil(t, m["null_value"])
}

func TestLoadSequenceOfMappings(t *testing.T) {
	yaml := `
- name: Alice
  age: 30
- name: Bob
  age: 25
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)
	expected := []any{
		map[string]any{"name": "Alice", "age": float64(30)},
		map[string]any{"name": "Bob", "age": float64(25)},
	}
	assert.Equal(t, expected, data)
}

func TestLoadMappingWithSequenceValues(t *testing.T) {
	yaml := `
fruits:
  - apple
  - banana
colors:
  - red
  - blue
`
	data, err := yamlstar.Load(yaml)
	require.NoError(t, err)
	expected := map[string]any{
		"fruits": []any{"apple", "banana"},
		"colors": []any{"red", "blue"},
	}
	assert.Equal(t, expected, data)
}

func TestLoadAllSingleDocument(t *testing.T) {
	data, err := yamlstar.LoadAll("hello")
	require.NoError(t, err)
	assert.Equal(t, []any{"hello"}, data)
}

func TestLoadAllMultipleDocuments(t *testing.T) {
	yaml := `---
doc1
---
doc2
---
doc3`
	data, err := yamlstar.LoadAll(yaml)
	require.NoError(t, err)
	assert.Equal(t, []any{"doc1", "doc2", "doc3"}, data)
}

func TestLoadAllWithExplicitMarkers(t *testing.T) {
	yaml := `---
a: 1
...
---
b: 2
...`
	data, err := yamlstar.LoadAll(yaml)
	require.NoError(t, err)
	expected := []any{
		map[string]any{"a": float64(1)},
		map[string]any{"b": float64(2)},
	}
	assert.Equal(t, expected, data)
}

func TestLibVersion(t *testing.T) {
	version, err := yamlstar.LibVersion()
	require.NoError(t, err)
	assert.NotEmpty(t, version)
}

func TestErrorHandlingMalformedYAML(t *testing.T) {
	// Unclosed quote is truly malformed
	_, err := yamlstar.Load(`key: "unclosed`)
	assert.Error(t, err)
}

func TestEmptyDocument(t *testing.T) {
	data, err := yamlstar.Load("")
	require.NoError(t, err)
	assert.Nil(t, data)
}

func TestWhitespaceOnly(t *testing.T) {
	data, err := yamlstar.Load("   \n  \n  ")
	require.NoError(t, err)
	assert.Nil(t, data)
}

func TestQuotedStrings(t *testing.T) {
	data1, err := yamlstar.Load("'hello world'")
	require.NoError(t, err)
	assert.Equal(t, "hello world", data1)

	data2, err := yamlstar.Load(`"hello world"`)
	require.NoError(t, err)
	assert.Equal(t, "hello world", data2)
}

func TestPackageVersion(t *testing.T) {
	assert.NotEmpty(t, yamlstar.Version)
}

File: csharp/test/YAMLStarTest.cs
using Xunit;
using YAMLStar;
using System.Text.Json;

namespace YAMLStar.Tests;

public class YAMLStarTest
{
    [Fact]
    public void TestRuntimeInitialization()
    {
        var ys = new YAMLStar();
        Assert.NotNull(ys);
        ys.Dispose();
    }

    [Fact]
    public void TestLoadSimpleScalar()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("hello");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal("hello", element.GetString());
    }

    [Fact]
    public void TestLoadInteger()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("42");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(42, element.GetInt32());
    }

    [Fact]
    public void TestLoadFloat()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("3.14");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(3.14, element.GetDouble(), precision: 2);
    }

    [Fact]
    public void TestLoadBooleanTrue()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("true");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.True(element.GetBoolean());
    }

    [Fact]
    public void TestLoadBooleanFalse()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("false");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.False(element.GetBoolean());
    }

    [Fact]
    public void TestLoadNull()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("null");
        if (result != null)
        {
            var element = Assert.IsType<JsonElement>(result);
            Assert.Equal(JsonValueKind.Null, element.ValueKind);
        }
        else
        {
            Assert.Null(result);
        }
    }

    [Fact]
    public void TestLoadSimpleMapping()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("key: value");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(JsonValueKind.Object, element.ValueKind);
        Assert.Equal("value", element.GetProperty("key").GetString());
    }

    [Fact]
    public void TestLoadNestedMapping()
    {
        using var ys = new YAMLStar();
        var yaml = @"
outer:
  inner: value
";
        var result = ys.Load(yaml);
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal("value", element.GetProperty("outer").GetProperty("inner").GetString());
    }

    [Fact]
    public void TestLoadSimpleSequence()
    {
        using var ys = new YAMLStar();
        var yaml = @"
- item1
- item2
- item3
";
        var result = ys.Load(yaml);
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(JsonValueKind.Array, element.ValueKind);
        Assert.Equal(3, element.GetArrayLength());
        Assert.Equal("item1", element[0].GetString());
        Assert.Equal("item2", element[1].GetString());
        Assert.Equal("item3", element[2].GetString());
    }

    [Fact]
    public void TestLoadFlowSequence()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("[a, b, c]");
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(JsonValueKind.Array, element.ValueKind);
        Assert.Equal(3, element.GetArrayLength());
    }

    [Fact]
    public void TestLoadAll()
    {
        using var ys = new YAMLStar();
        var yaml = @"---
doc1
---
doc2
---
doc3";
        var result = ys.LoadAll(yaml);
        Assert.NotNull(result);
        var element = Assert.IsType<JsonElement>(result);
        Assert.Equal(JsonValueKind.Array, element.ValueKind);
        Assert.Equal(3, element.GetArrayLength());
        Assert.Equal("doc1", element[0].GetString());
        Assert.Equal("doc2", element[1].GetString());
        Assert.Equal("doc3", element[2].GetString());
    }

    [Fact]
    public void TestVersion()
    {
        using var ys = new YAMLStar();
        var version = ys.Version();
        Assert.NotNull(version);
        Assert.NotEmpty(version);
    }

    [Fact]
    public void TestErrorHandling()
    {
        using var ys = new YAMLStar();
        Assert.Throws<YAMLStarException>(() => ys.Load("key: \"unclosed"));
    }

    [Fact]
    public void TestDispose()
    {
        var ys = new YAMLStar();
        ys.Dispose();

        // Should throw when disposed
        Assert.Throws<ObjectDisposedException>(() => ys.Load("test: value"));
    }

    [Fact]
    public void TestEmptyDocument()
    {
        using var ys = new YAMLStar();
        var result = ys.Load("");
        if (result != null)
        {
            var element = Assert.IsType<JsonElement>(result);
            Assert.Equal(JsonValueKind.Null, element.ValueKind);
        }
        else
        {
            Assert.Null(result);
        }
    }

    [Fact]
    public void TestQuotedStrings()
    {
        using var ys = new YAMLStar();
        var result1 = ys.Load("'hello world'");
        Assert.NotNull(result1);
        var element1 = Assert.IsType<JsonElement>(result1);
        Assert.Equal("hello world", element1.GetString());

        var result2 = ys.Load("\"hello world\"");
        Assert.NotNull(result2);
        var element2 = Assert.IsType<JsonElement>(result2);
        Assert.Equal("hello world", element2.GetString());
    }
}
